### PR TITLE
tests: new spread log utils

### PR DESCRIPTION
**Pull request by sergiocazzolato**
The log utilities from snapd-testing-tools have been updated
1. log-helper: has enums instead of constants
2. log-parser: does not offer --print-details functions anymore, and the complexity was reduced. Also many other minor improvements have been implemented. 
3. log-analyzer: minor changes
4. minor fix in remote.pull to make it work in ubuntu noble with nested tests

### CODE REVIEW

File: tests/lib/external/snapd-testing-tools/CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -6,7 +6,7 @@ In the interest of fostering an open and welcoming environment, we as
 contributors and maintainers pledge to making participation in our project and
 our community a harassment-free experience for everyone, regardless of age, body
 size, disability, ethnicity, sex characteristics, gender identity and expression,
-level of experience, education, socio-economic status, nationality, personal
+level of experience, education, socioeconomic status, nationality, personal
```
Comment (sergiocazzolato) [inline tests/lib/external/snapd-testing-tools/CODE_OF_CONDUCT.md:9]: @bboozzoo wrote: wonder where this change comes from

File: tests/lib/external/snapd-testing-tools/utils/log-parser [hunk 2]
```diff
@@ -310,47 +254,38 @@ class LogReader:
                 prepared_log.append(item)
             else:
                 prepared_log.append(item.__dict__())
-        with open(filepath, 'w') as json_file:
+        with open(filepath, "w") as json_file:
             json.dump(prepared_log, json_file, indent=4)
 
     def _next_line(self):
```
Comment (sergiocazzolato) [inline tests/lib/external/snapd-testing-tools/utils/log-parser:260]: @zyga wrote: Why is this not just a iter(self.lines) somewhere? Please remove this.

File: tests/lib/external/snapd-testing-tools/utils/log-parser [hunk 3]
```diff
@@ -160,17 +126,21 @@ class Rule:
...
 
         try:
             re.compile(self.pattern)
         except re.error as err:
-            raise ValueError("Error: pattern '{}' cannot be compiled: {}".format(pattern, err))
+            raise ValueError(
+                "Error: pattern '{}' cannot be compiled: {}".format(self.pattern, err)
+            )
 
     def filter(self, lines):
```
Comment (sergiocazzolato) [inline tests/lib/external/snapd-testing-tools/utils/log-parser:145]: @zyga wrote: This should re-use a compiled pattern stored as an instance attribute.

File: tests/lib/external/snapd-testing-tools/utils/log-filter [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+                "Error: Rule operation '{}' not in supported list: {}".format(
+                    rule_operation, SUPPORTED_RULES
+                )
+            )
+
+        if operation == rule_operation:
+            patterns.append(rule_pattern)
+
+    for pattern in patterns:
+        regex_list.append(re.compile(pattern))
```
Comment (sergiocazzolato) [inline tests/lib/external/snapd-testing-tools/utils/log-filter:59]: @zyga wrote: There's no need to create another list. Construct the regex_list in the logic above.

File: tests/lib/external/snapd-testing-tools/utils/log-parser [hunk 5]
```diff
@@ -68,12 +34,12 @@ class Action:
 
     def __dict__(self):
```
Comment (sergiocazzolato) [inline tests/lib/external/snapd-testing-tools/utils/log-parser:35]: @zyga wrote:
I've left a remark in my commit message but we should not invent new dunder methods or change the behavior of existing methods. For as long as I remember dict is a special attribute that is a dictionary of the internals of the value. It is not a method and its content directly corresponds to the state of the object.

I've renamed this to _some_dict, please check if this is what you expected.

File: tests/lib/external/snapd-testing-tools/utils/log-parser [hunk 6]
```diff
@@ -82,9 +48,10 @@ class Result:
     The results can be: Successful, failed and aborted
     """
 
-    def __init__(self, result_type, level, stage, number, date, time,
-                 detail, source_line):
-        self.type = RESULT
+    def __init__(
```
Comment (sergiocazzolato) [inline tests/lib/external/snapd-testing-tools/utils/log-parser:51]: @zyga wrote:
There's a lot of string arguments that are somewhat confusing. What is level? What is number?

Please document those.

File: tests/lib/external/snapd-testing-tools/utils/log-parser [hunk 7]
```diff
@@ -96,34 +63,33 @@ class Result:
...
-    Info represents the extra tasks information which is included in the
-    spread log. The info can be: Error, Debug and Warning
+    Info represents the tasks status information which is included
+    in the spread log. The info can be: Error, Debug and Warning.
     """
 
-    def __init__(self, info_type, verb, task, extra, date, time,
-                 detail, source_line):
-        self.type = INFO
+    def __init__(self, info_type, verb, task, extra, date, time, detail, source_line):
```
Comment (sergiocazzolato) [inline tests/lib/external/snapd-testing-tools/utils/log-parser:91]: @zyga wrote: There's more confusing undocumented arguments here: what is extra for?

Please document it.

File: tests/lib/external/snapd-testing-tools/utils/log-parser [hunk 8]
```diff
@@ -160,17 +126,21 @@ class Rule:
...
-            raise ValueError("Error: Rule '{}' does not follow the KEY=PATTERN format".format(rule))
+            raise ValueError(
+                "Error: Rule '{}' does not follow the KEY=PATTERN format".format(rule)
+            )
 
         self.key = parts[0]
         self.pattern = parts[1]
 
         try:
             re.compile(self.pattern)
```
Comment (sergiocazzolato) [inline tests/lib/external/snapd-testing-tools/utils/log-parser:139]: @zyga wrote:
You are throwing away the result of the compilation. Just store it as an instance attribute and remove all the pointless re-compilation form the filter method.

File: tests/lib/external/snapd-testing-tools/utils/log_helper.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
+"""
+This library provides a set of functions that can be used to
+process the spread output.
+"""
+
+import re
+
```
Comment (sergiocazzolato) [inline tests/lib/external/snapd-testing-tools/utils/log_helper.py:7]: @zyga wrote:
All of those should be proper enums, depending on the type of the interaction you need either a pure enum, a string enum or something else. A set of free-flying strings defined and used in a random places makes it hard to understand what each function can really process.

File: tests/lib/external/snapd-testing-tools/utils/log-filter [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+
+def print_line(line: str):
+    if not line:
+        print()
+    else:
+        print(line.strip())
+
+
+def write_line(line: str, to_stream: io.TextIOWrapper):
```
Comment (sergiocazzolato) [inline tests/lib/external/snapd-testing-tools/utils/log-filter:25]: @zyga wrote:
Please combine print_line and write_line, we have print(..) optional argument to do both.

File: tests/lib/external/snapd-testing-tools/CODE_OF_CONDUCT.md (hunk 1 above)
Comment (sergiocazzolato) [reply tests/lib/external/snapd-testing-tools/CODE_OF_CONDUCT.md:9]: the spellcheck required the change

File: tests/lib/external/snapd-testing-tools/utils/log-parser (hunk 5 above)
Comment (sergiocazzolato) [reply tests/lib/external/snapd-testing-tools/utils/log-parser:35]: done

**Comment by sergiocazzolato** [comment]
@zyga Annotations added. Also I included the comments from the previous pr.

File: tests/lib/external/snapd-testing-tools/utils/log_helper.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+from enum import Enum
+
+
+class ListedEnum(Enum):
+    @classmethod
+    def list(cls):
+        return list(map(lambda c: c.value, cls))
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log_helper.py:14]: This should return `cls.__members__.values()`.

Add a type annotation please.

**Review by zyga — APPROVED** [review]
can you make some more changes:

```diff
diff --git a/tests/lib/external/snapd-testing-tools/utils/log_helper.py b/tests/lib/external/snapd-testing-tools/utils/log_helper.py
index 1c95ba731e..030d43f026 100644
--- a/tests/lib/external/snapd-testing-tools/utils/log_helper.py
+++ b/tests/lib/external/snapd-testing-tools/utils/log_helper.py
@@ -10,8 +10,8 @@ from enum import Enum
 
 class ListedEnum(Enum):
     @classmethod
-    def list(cls):
-        return list(map(lambda c: c.value, cls))
+    def list(cls) -> list[str]:
+        return [str(x) for x in cls.__members__.values()]
 
 
 class ExecutionPhase(ListedEnum):
@@ -65,11 +65,11 @@ OPERATIONS = (
 START_LINE = ".*Project content is packed for delivery.*"
 
 
-def _match_date(date) -> bool:
+def _match_date(date: str) -> bool:
     return re.match(r"\d{4}-\d{2}-\d{2}", date) is not None
 
 
-def _match_time(time) -> bool:
+def _match_time(time: str) -> bool:
     return re.match(r"\d{2}:\d{2}:\d{2}", time) is not None
 
```

Please apply this to `log_helper.py`
Remember that by default mypy uses gradual typing, to be sure we're not missing anything pass `--strict` each time


We should also look at `log-parser:
```
zyga@novigrad:~/Canonical/snapd/tests/lib/external/snapd-testing-tools/utils$ mypy --strict log-parser
log-parser:24: error: Function is missing a type annotation  [no-untyped-def]
log-parser:32: error: Function is missing a type annotation  [no-untyped-def]
log-parser:35: error: Function is missing a return type annotation  [no-untyped-def]
log-parser:51: error: Function is missing a type annotation  [no-untyped-def]
log-parser:64: error: Function is missing a type annotation  [no-untyped-def]
log-parser:69: error: Function is missing a return type annotation  [no-untyped-def]
log-parser:91: error: Function is missing a type annotation  [no-untyped-def]
log-parser:102: error: Function is missing a type annotation  [no-untyped-def]
log-parser:107: error: Function is missing a return type annotation  [no-untyped-def]
log-parser:145: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
log-parser:176: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
log-parser:182: error: Function is missing a type annotation  [no-untyped-def]
log-parser:211: error: Function is missing a type annotation  [no-untyped-def]
log-parser:247: error: Function is missing a type annotation  [no-untyped-def]
log-parser:250: error: Missing type parameters for generic type "dict"  [type-arg]
log-parser:295: error: Call to untyped function "_get_phase" in typed context  [no-untyped-call]
log-parser:302: error: Call to untyped function "_get_info" in typed context  [no-untyped-call]
log-parser:310: error: Call to untyped function "_get_action" in typed context  [no-untyped-call]
log-parser:317: error: Call to untyped function "_get_result" in typed context  [no-untyped-call]
log-parser:337: error: Function is missing a type annotation  [no-untyped-def]
log-parser:365: error: Function is missing a type annotation  [no-untyped-def]
log-parser:402: error: Call to untyped function "_get_detail" in typed context  [no-untyped-call]
log-parser:404: error: Call to untyped function "Info" in typed context  [no-untyped-call]
log-parser:406: error: Function is missing a type annotation  [no-untyped-def]
log-parser:420: error: Call to untyped function "_get_detail" in typed context  [no-untyped-call]
log-parser:422: error: Call to untyped function "Result" in typed context  [no-untyped-call]
log-parser:424: error: Function is missing a type annotation  [no-untyped-def]
log-parser:432: error: Call to untyped function "Phase" in typed context  [no-untyped-call]
log-parser:434: error: Function is missing a type annotation  [no-untyped-def]
log-parser:442: error: Argument 2 to "Action" has incompatible type "None"; expected "str"  [arg-type]
Found 30 errors in 1 file (checked 1 source file)
```

With the changes I posted earlier, it was 100% clean so perhaps we can open up meld and restore them as a follow-up?

File: tests/lib/external/snapd-testing-tools/utils/log_helper.py (hunk 11 above)
Comment (bboozzoo) [reply tests/lib/external/snapd-testing-tools/utils/log_helper.py:14]: I think this only returns the mappings, so you'd have to do `[v.value for v in cls.__members__.values()]` to get the list of actual string values associated with each member

**Review by bboozzoo — APPROVED** [review]
LGTM

**Pull request merged**